upup/pkg/fi/cloudup/terraformWriter/writer.go

```diff
@@ -88,6 +92,28 @@ func (t *TerraformWriter) AddFileBytes(resourceType string, resourceName string,
 	return LiteralFunctionExpression(fn, path), nil
 }
 
+func (t *TerraformWriter) EnsureTerraformProvider(tfProvider *TerraformProvider) error {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	if t.Providers == nil {
+		t.Providers = make(map[string]*TerraformProvider)
+	}
+
+	key := tfProvider.Name
+
+	existing := t.Providers[key]
+	if existing != nil {
+		if reflect.DeepEqual(tfProvider, existing) {
+			// already exists and matches
+			return nil
+		}
+		return fmt.Errorf("attempt to add different tfProvider with key %q", key)
```

Comment:
Consider panic() instead?

Actually, consider having this take name and arguments as parameters, then return an interned `*TerraformProvider`.

Reply:
Changed to panic.

I was thinking along similar lines, it would be nice if the vfs package did not depend on the terraform package, and refactoring EnsureProvider to take the parameters would be good here (after this PR, we don't need to expose the TerraformProvider function I don't think).  But that is a much bigger refactor, and I wanted to keep this one smallish.  I can give it a go though, and we can see what it looks like in a separate PR (it'll build on this one anyway!)